### PR TITLE
feat: implement cons@k evaluation

### LINKED ISSUE #541: Support cons@k in eval

### PR DESCRIPTION
**Pull request by RayenTian**
# What does this PR do ?

Add support for cons@k metric in evaluation module to improve performance measurement.
To be more specific:
        1. Add extracted_answer to EnvironmentReturn.metadata.
        2. Implement the function of selecting the majority of answers among k_value answers.
        3. Enumerate all combinations in C(k, n) and calculate the corresponding scores, and finally get an unbiased estimate of cons@k.

# Issues
Closes #541 

# Test Result
When k_value is set to 1, the values of pass@k and cons@k should be equal. By adding some print statements and running the following command:
```bash
uv run python examples/run_eval.py eval.metric="cons@k"
```

the results are as follows:
```
pass@k: 0.1(3.0/30)
cons@k: 0.1(3.0/30)

============================================================
model_name='Qwen2.5-Math-1.5B-Instruct' dataset_name='aime2024'
max_new_tokens=2048 temperature=0.0 top_p=1.0 top_k=-1

metric='cons@k' k_value=1 num_tests_per_prompt=1

score=0.1000 (3.0/30)
============================================================
```

For k_value=3 and num_tests_per_prompt=5, we set temperature=0.7. By adding some print statements and running the command:
```bash
uv run python examples/run_eval.py \
    generation.temperature=0.7 \
    eval.metric="cons@k" \
    eval.k_value=3 \
    eval.num_tests_per_prompt=5 
```
we obtain the following results:
```
============================================================
model_name='Qwen2.5-Math-1.5B-Instruct' dataset_name='aime2024'
max_new_tokens=2048 temperature=0.7 top_p=1.0 top_k=-1

metric='cons@k' k_value=3 num_tests_per_prompt=5

score=0.0833 (2.5/30)
============================================================
```

More intermediate results are provided in the [test_3_5.log](https://github.com/user-attachments/files/21280947/test_3_5.log) file, where you can see the calculation process and verify its correctness.

# Usage
1. Modify `examples/configs/evals/eval.yaml` as you need. Note that the metric field need to be set to cons@k.
2. Start up the eval process.
```python
uv run python examples/run_eval.py generation.temperature=0.7 eval.metric="cons@k" eval.k_value=3 eval.nu
m_tests_per_prompt=5 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](/NVIDIA-NeMo/RL/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you run the unit tests and functional tests locally? Visit our [Testing Guide](/NVIDIA-NeMo/RL/blob/main/docs/testing.md) for how to run tests
- [ ] Did you add or update any necessary documentation? Visit our [Document Development Guide](/NVIDIA-NeMo/RL/blob/main/docs/documentation.md) for how to write, build and test the docs.

# Additional Information

1. The cons@k metric uses an answer voting method to select the final answer. However, how to handle tie votes remains an open question. Currently, we choose the first generated answer in case of ties, but whether this approach is reasonable requires further discussion.
2. The math_metric does not allow passing an empty function, so the current operation is to obtain the corresponding predict_answer according to the scoring logic of math_metric.

### CODE REVIEW

File: 3rdparty/NeMo-workspace/NeMo
Comment (yuki-666) [inline 3rdparty/NeMo-workspace/NeMo:1]: Is it submitted accidentally? Eval seems not need this.

Comment (yuki-97) [inline 3rdparty/NeMo-workspace/NeMo:1]: Is it submitted accidentally? Eval seems not need this.

File: examples/configs/evals/eval.yaml
Comment (yuki-97) [inline examples/configs/evals/eval.yaml:1]: Can you revert the changes in the example config? Otherwise functional test may fail.

File: examples/configs/evals/eval.yaml [hunk 1]
```diff
@@ -1,14 +1,14 @@
 # Evaluation Configuration
 eval:
-  metric: "pass@k"
-  num_tests_per_prompt: 1 # every prompt will be tested num_tests_per_prompt times and use the average score as the final score
+  metric: "cons@k"
```
Comment (yuki-97) [inline examples/configs/evals/eval.yaml]:
```suggestion
  metric: "pass@k" # pass@k and cons@k are supported
```

File: nemo_rl/evals/eval.py [hunk 2]
```diff
@@ -86,24 +88,21 @@ def setup(
...
         assert temperature > 0 and top_k != 1, (
             "temperature > 0 and top_k != 1 are required for multiple samples"
         )
 
-    assert pass_k_value >= 1, (
-        "pass_k_value must be greater than or equal to 1 for pass@k metric"
-    )
-    assert num_tests_per_prompt >= pass_k_value, (
-        "num_tests_per_prompt must be greater than or equal to pass_k_value for pass@k metric"
+    assert k_value >= 1, "k_value must be greater than or equal to 1 for pass@k metric"
```
Comment (yuki-97) [inline nemo_rl/evals/eval.py]:
```suggestion
    assert k_value >= 1, "k_value must be greater than or equal to 1"
```
`num_tests_per_prompt` is also.

File: nemo_rl/environments/math_environment.py [hunk 3]
```diff
@@ -68,53 +69,83 @@ def __init__(self) -> None:
...
+                    assert extracted_answer is not None
+                    assert len(extracted_answer) == 2
+                    extracted_prediction, extracted_gold = extracted_answer
+                    for pred in extracted_prediction:
+                        if any(grader.verify(gold, pred) for gold in extracted_gold):
+                            extracted_answers.append(pred)
+                            break
+                    else:
+                        extracted_answers.append(None)
             except Exception:
```
Comment (yuki-97) [inline nemo_rl/environments/math_environment.py]: Can we only keep this and remove line 102~104?

File: nemo_rl/environments/math_environment.py [hunk 4]
```diff
@@ -250,15 +313,21 @@ def step(  # type: ignore[override]
         # create a tensor of rewards and done flags
         rewards = torch.tensor(results).cpu()
         done = torch.ones_like(rewards).cpu()
-
         next_stop_strings = [None] * len(message_log_batch)
 
+        info = (
+            [{"extracted_answer": answer} for answer in extracted_answers_list]
```
Comment (yuki-97) [inline nemo_rl/environments/math_environment.py]: Feels too many duplicated keys `extracted_answer`. Any reason not using `info = {"extracted_answer": extracted_answers_list} if return_extracted_answer else None`?

File: nemo_rl/evals/eval.py [hunk 5]
```diff
@@ -191,6 +190,80 @@ def eval_single_chunk(n: int, c: int, k: int) -> float:
...
+
+    Returns:
+        cons_k_score: float
+    """
+
+    def majority_vote(answers: list[str | None]) -> str | None:
+        """Find the most common answer in the list of answers."""
+        if not answers:
+            return None
+        # To fix@rayentian: How to deal with the case that there are multiple most common answers? Now we just return the first one.
```
Comment (yuki-97) [inline nemo_rl/evals/eval.py:219]: How does others do this? (other eval repo or paper)
If there is no reference, how about average on all most common answers?

**Review by yuki-666 — COMMENTED** [review]
Thanks @RayenTian ! Left some comments.

And can you add a test result which shows the cons@k implementation is correct and paste it in the PR?

File: nemo_rl/environments/math_environment.py (hunk 4 above)
Comment (RayenTian) [reply nemo_rl/environments/math_environment.py]: The Info field is designed as list[Optional[dict]] to maintain consistency with the formats of EnvironmentReturn.observations (list[dict[str, str]]) and EnvironmentReturn.metadata (list[Optional[dict]]).
And considering that more information may be added in the future, an extracted_answer key has been included.

File: nemo_rl/evals/eval.py (hunk 5 above)
Comment (RayenTian) [reply nemo_rl/evals/eval.py:219]: I haven’t found a particularly good explanation of this situation in open-source code or papers. For some math problems, the possible answers can vary greatly, so taking their average may result in an implausible or even impossible answer, even if the correct answer is among the original candidate responses. Moreover, averaging requires recomputation based on the LLM's outputs, which I believe is not an appropriate approach for evaluation.

File: nemo_rl/environments/math_environment.py (hunk 3 above)
Comment (RayenTian) [reply nemo_rl/environments/math_environment.py]: Done.

**Comment by RayenTian** [comment]
> Thanks @RayenTian ! Left some comments.
> 
> And can you add a test result which shows the cons@k implementation is correct and paste it in the PR?

When k_value is set to 1, the values of pass@k and cons@k should be equal. By adding some print statements and running the following command:
```bash
uv run python examples/run_eval.py eval.metric="cons@k"
```

the results are as follows:
```
pass@k: 0.1(3.0/30)
cons@k: 0.1(3.0/30)

============================================================
model_name='Qwen2.5-Math-1.5B-Instruct' dataset_name='aime2024'
max_new_tokens=2048 temperature=0.0 top_p=1.0 top_k=-1

metric='cons@k' k_value=1 num_tests_per_prompt=1

score=0.1000 (3.0/30)
============================================================
```

For k_value=3 and num_tests_per_prompt=5, we set temperature=0.7. By adding some print statements and running the command:
```bash
uv run python examples/run_eval.py \
    generation.temperature=0.7 \
    eval.metric="cons@k" \
    eval.k_value=3 \
    eval.num_tests_per_prompt=5 
```
we obtain the following results:
============================================================
model_name='Qwen2.5-Math-1.5B-Instruct' dataset_name='aime2024'
max_new_tokens=2048 temperature=0.7 top_p=1.0 top_k=-1

metric='cons@k' k_value=3 num_tests_per_prompt=5

score=0.0833 (2.5/30)
============================================================


More intermediate results are provided in the [test_3_5.log](https://github.com/user-attachments/files/21280947/test_3_5.log) file, where you can see the calculation process and verify its correctness.

File: nemo_rl/environments/math_environment.py (hunk 4 above)
Comment (yuki-97) [reply nemo_rl/environments/math_environment.py]: Got it, thanks for clarifying!

File: nemo_rl/evals/eval.py (hunk 5 above)
Comment (yuki-97) [reply nemo_rl/evals/eval.py:219]: Got it, let's keep your current way for now.

**Review by yuki-666 — APPROVED** [review]
Thanks @RayenTian , LGTM!
Let's wait merging it after v0.3 is released.

**Review by joyang-nv — APPROVED** [review]
I reviewed more on rollouts and eval. Changes looks good to me.

File: nemo_rl/environments/interfaces.py [hunk 6]
```diff
@@ -34,13 +34,15 @@ class EnvironmentReturn(NamedTuple):
...
     terminateds: whether the episode ended this turn.
+    info: any additional information from the environment.
     """
 
     observations: list[dict[str, str]]
     metadata: list[Optional[dict]]
     next_stop_strings: list[list[str] | None] | list[None]
     rewards: Tensor
     terminateds: Tensor
+    info: list[Optional[dict]]
```
Comment (terrykong) [inline nemo_rl/environments/interfaces.py]: should we maybe just name this `answers` or something narrow? it makes it much easier for the IDEs to figure out what's in this environment return and helps people know if they should return something there or not

**Review by terrykong — COMMENTED** [review]
looks good to me aside from my one comment

thanks for making nemo rl better!

File: nemo_rl/environments/interfaces.py (hunk 6 above)
Comment (RayenTian) [reply nemo_rl/environments/interfaces.py]: Thank you for the suggestion! I have updated the name to answers as recommended.

**Review by terrykong — DISMISSED** [review]

File: nemo_rl/evals/eval.py [hunk 7]
```diff
@@ -321,6 +401,6 @@ def _print_results(
     print("\n" + "=" * 60)
     print(f"{model_name=} {dataset_name=}")
     print(f"{max_new_tokens=} {temperature=} {top_p=} {top_k=}\n")
-    print(f"{metric=} {pass_k_value=} {num_tests_per_prompt=}\n")
+    print(f"{metric=} {k_value=} {num_tests_per_prompt=}\n")
```
Comment (yuki-97) [inline nemo_rl/evals/eval.py]:
```suggestion
    print(f"metric={metric[:-1]}{k_value} {num_tests_per_prompt=}\n")
```

A minor thing that can you rebase main and update the `## Example Evaluation Output` in `docs/guides/eval.md`?

https://github.com/NVIDIA-NeMo/RL/blob/main/docs/guides/eval.md?plain=1#L84
from
```
metric='pass@k' pass_k_value=1 num_tests_per_prompt=1
```
to
```
metric=pass@1 num_tests_per_prompt=1
```

Comment (yuki-97) [reply nemo_rl/evals/eval.py]: hmm, I mean `metric={metric[:-1]}{k_value}` and `metric=pass@1`, instead of `metric={metric}@{k_value}` and `metric=pass@k@1`.

**Review by yuki-666 — APPROVED** [review]

**Comment by RayenTian** [comment]
After the rebase, the existing comments and approvals have become invalid, so I kindly ask everyone to review and approve once more.

**Review by yuki-666 — APPROVED** [review]

**Review by terrykong — APPROVED** [review]

**Review by yuki-666 — DISMISSED** [review]

**Review by parthchadha — APPROVED** [review]

**Comment by RayenTian** [comment]
PR merge failed the CI tests because the recently merged code introduced a code_environment.py file, and its interface was unchanged. I have now updated the code accordingly. Could you please review it again? Thank you!

File: nemo_rl/environments/code_environment.py [hunk 8]
```diff
@@ -236,14 +237,19 @@ def step(
             terminated_tensor = torch.zeros(len(terminateds), dtype=torch.bool)
         rewards_tensor = torch.zeros_like(terminated_tensor, dtype=torch.float32)
 
-        next_stop_strings = [["</code>"]] * len(message_log_batch)
+        next_stop_strings = [None] * len(message_log_batch)
```
Comment (yuki-97) [inline nemo_rl/environments/code_environment.py]: Can you revert the change of `next_stop_strings`? I think there's no reason to change it.

File: nemo_rl/environments/code_environment.py [hunk 9]
```diff
@@ -236,14 +237,19 @@ def step(
...
+        if return_extracted_answer:
```
Comment (yuki-97) [inline nemo_rl/environments/code_environment.py]: Can you use something like `assert return_extracted_answer == False` and add a comment that return_extracted_answer is not supported here instead of return a mock data?

**Review by yuki-666 — APPROVED** [review]

**Review by yuki-666 — APPROVED** [review]

File: nemo_rl/evals/eval.py [hunk 10]
```diff
@@ -195,6 +194,80 @@ def eval_single_chunk(n: int, c: int, k: int) -> float:
     return pass_k_score
 
 
+def eval_cons_k(
```
Comment (terrykong) [inline nemo_rl/evals/eval.py:197]: @RayenTian could you write a unit test for this?

Comment (yuki-97) [reply nemo_rl/evals/eval.py:197]: it would be helpful if you can also help to add a unit test for `eval_pass_k`.

**Review by terrykong — APPROVED** [review]
